src/classes/RD2_OpportunityEvaluationService.cls

```diff
@@ -455,6 +471,9 @@ public inherited sharing class RD2_OpportunityEvaluationService {
         if (isBatchContext) {
             return !rdSettings.isDisableAllInstallments;
 
+        } else if (rd.isNew() && triggerAction == TDTM_Runnable.Action.AfterUpdate) {
```

Comment:
I'm just wondering if it's better to control the flow of calling the service to the [RD2_RecurringDonations_TDTM](https://github.com/SalesforceFoundation/NPSP/blob/master/src/classes/RD2_RecurringDonations_TDTM.cls#L380) class rather than here. Assuming it's possible to validate the same `isNew` condition in that class, we could just not call the Queueable from the trigger on an AfterUpdate context. Evaluating `isNew` may not be feasible in the TDTM class too, but I wanted to throw it out there. If I follow the logic correctly, `isNew` is determined by there being no related Opps. If true, that likely can't be determined in the TDTM class anyway.

Reply:
RD2_RecurringDonations_TDTM needs to call RD2_OpportunityEvaluationService from both AfterInsert and AfterUpdate.  In the case of a potentially changed schedule, it needs to take a look at the existing Opportunities and take appropriate action.

You're right, isNew() is based on whether there are related opportunities.

My conclusion was that the approach least likely to have unintended side effects was to isolate the narrow set of circumstances that produced the bug, and that is the combination of a Recurring Donation that RD2_RecurringDonation determines is new based upon no related Opportunities and a call from RD2_RecurringDonations_TDTM with an AfterUpdate action type.

Reply:
Makes sense. This also confirms that my initial guess on a cause was way off. I had originally thought that it was calling the queueable only when it thought a new Opp was needed, but that was wrong. The queueable itself makes the proper determination rather than the trigger. As you've identified though, it's getting hit from both an after insert and an after update.

What would happen in the following scenario (not that I think this is likely)?
- A new RD is created and a new Opp is created with it
- The open Opp is deleted (manually or otherwise)

Would a new opp be created by the trigger, or would that happen when the nightly job runs?

Reply:
Related to the initial comment and the reply above, if the Opp eval service is called after the update or new RDs in the case Schedule changes, should we actually skip the Opp eval service call after the insert and call it only once within the transaction, ie `afterUpdate`? If so, then we would keep the trigger actions within the trigger scope and the Opp eval queueable and the service would not need to know of the trigger context. Thoughts?

Reply:
I'd like us to meet to discuss this offline.  I'm concerned that we are about to initiate a major change to fix an isolated bug.